src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -0,0 +1,134 @@
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            if (_options.IncludeRequestBody)
+            {
+                httpContext.Request.EnableBuffering();
+            }
+
+            try
+            {
+                await _next(httpContext);
+            }
+            finally
+            {
+                stopwatch.Stop();
+                await TrackRequestAsync(httpContext, stopwatch.Elapsed);
+            }
+        }
+
+        private async Task TrackRequestAsync(HttpContext httpContext, TimeSpan duration)
+        {
+            try
+            {
+                IDictionary<string, StringValues> headers = new Dictionary<string, StringValues>();
+                if (_options.IncludeRequestHeaders)
+                {
+                    IDictionary<string, StringValues> sanitizedHeaders = SanitizeRequestHeaders(httpContext.Request.Headers);
+                    if (sanitizedHeaders != null)
+                    {
+                        headers = sanitizedHeaders;
+                    }
+                }
+
+                IDictionary<string, StringValues> body = new Dictionary<string, StringValues>();
+                if (_options.IncludeRequestBody)
+                {
+                    IDictionary<string, StringValues> sanitizedBody = await SanitizeRequestBodyAsync(httpContext.Request.Body);
+                    if (sanitizedBody != null)
+                    {
+                        body = sanitizedBody;
+                    }
+                }
+
+                Dictionary<string, object> logContext = headers.Concat(body).ToDictionary(kv => kv.Key, kv => (object) kv.Value);
+                _logger.LogRequest(httpContext.Request, httpContext.Response, duration, logContext);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogCritical(exception, "Failed to track request");
+            }
+        }
+
+        /// <summary>
+        /// Extracts information from the given HTTP <paramref name="requestHeaders"/> to include in the request tracking context.
+        /// </summary>
+        /// <param name="requestHeaders">The headers of the current HTTP request.</param>
+        protected virtual IDictionary<string, StringValues> SanitizeRequestHeaders(IDictionary<string, StringValues> requestHeaders)
+        {
+            return requestHeaders.Where(header => _options.OmittedHeaderNames.Contains(header.Key) == false);
+        }
+
+        private static async Task<IDictionary<string, StringValues>> SanitizeRequestBodyAsync(Stream requestStream)
+        {
+            if (!requestStream.CanRead)
+            {
+                return new Dictionary<string, StringValues>();
+            }
+
+            if (requestStream.CanSeek)
+            {
+                if (requestStream.Position > 0)
+                {
+                    requestStream.Position = 0;
+                }
+            }
```

Comment:
I would say that we should not log the body if the stream doesn't support seeking ?
In that case (non seekable stream), you won't be able to reset it to it's initial position.

Reply:
That could be an extra fail-safe, yes, bus as I told here above, using `.EnabledBuffering()` will hold the request body in memory and will write it on disk when it exceeds a given threshold.

Reply:
On the other hand, the body can not be seekable, but could be set on position `0`.

Reply:
Voila, [changed it](https://github.com/arcus-azure/arcus.webapi/pull/140/files#diff-dd8dfb1d0fb4a65ce66a770f805d082bR124), OK for you? @fgheysels @tomkerkhove ?

Reply:
> On the other hand, the body can not be seekable, but could be set on position `0`.

AFAIK that won't work as you are "seeking" it to 0

Reply:
The position is only set to zero if the stream is seekable, so that's ok @tomkerkhove ?

Reply:
I was merely explaining why but didn't check the code